docs/regalloc-demo/demo.md

```diff
@@ -0,0 +1,363 @@
+```
+
+If you plan on doing development work on this checkout of ml-compiler-opt,
+use the `/ml-compiler-opt/requirements-ci.txt` requirements file to install
+some additional development dependencies.
+
+## Building Chromium
+
+To build Chromium, make sure you are in the `/chromium/src` directory and then
+run the following command to open a CLI text editor that will allow you to
+configure build settings:
+
+```bash
+gn args ./out/Release
+```
+
+Then, use the following configuration options:
+
+```
+is_official_build=true
+use_thin_lto=false
+is_cfi=false
+clang_embed_bitcode=true
+is_debug=false
+symbol_level=0
+enable_nacl=false
+```
+
+Immedaitely after closing the editor, `gn` will generate all of the files
+necessary so that `ninja` can execute all the necessary compilation steps.
+However, to extract a corpus for ML training, we also need a database of
+compilation commands. This can be obtained by running the following command:
+
+```
+gn gen ./out/Release --export-compile-commands
+```
+
+Then you can build Chromium with the `autoninja` utility:
+
+```bash
+autoninja -C ./out/Release
+```
+
+A full Chromium compile will take at least an hour on pretty well specced
+hardware (ie 96 thread work station) and much longer on lower specced
+hardware.
+
+TODO(boomanaiden154): Investigate the source of this assertion error.
+
+**Note:** If the build fails in the last couple steps tripping an assertion in
+the linker, you can safely ignore this. Preparing a corpus for ML training
+only requires the object files that get fed to the linker.
+
+## Building LLVM
+
+To build LLVM to train ML models, we first need to build TFLite and some
+dependencies so that we can embed it within LLVM to load and execute models
+on the fly during reinforcement learning. There is a script within this
+repository that clones and builds everything automatically and prepares
+a CMake cache file that can be passed to CMake during the LLVM build
+configuration. Running the script looks like this:
+
+```bash
+mkdir tflite
+cd tflite
+/ml-compiler-opt/buildbot/build_tflite.sh
+```
+
+This script should only take a couple minutes to execute as all the libraries
+that it pulls and builds are relatively small.
+
+Now, create a new folder to do an LLVM build and configure it using CMake:
+```bash
+mkdir llvm-build
+cd llvm-build
+cmake -G Ninja \
+  -DCMAKE_BUILD_TYPE=Release \
+  -DLLVM_ENABLE_PROJECTS="clang" \
+  -C /tflite/tflite.cmake \
+  /llvm-project/llvm
+```
+
+Now you can run the actual build with the following command:
+```bash
+cmake --build .
+```
+
+## ML training
+
+All of the following example commands assume you are working from within
+your checkout of the ml-compiler-opt repository:
+
+```bash
+cd /ml-compiler-opt
+```
+
+To start off training, we need to extract a per-object bitcode/compilation
+flag corpus. This can be done with the following script:
+
+```bash
+PYTHONPATH=$PYTHONPATH:. python3 compiler_opt/tools/extract_ir.py \
```

Comment:
is chrome of thinlto build?

Reply:
I'm currently not doing a ThinLTO build just because the patch that I have prepared currently is designed to work without ThinLTO. I need to write up a patch to add the necessary flags to extract a corpus using ThinLTO to do some validation for the Chromium patch, so I'll try and add a section on ThinLTO corpus extraction once that is ready.

Reply:
Just added info on thinLTO corpus extraction.